### PR TITLE
Add `IEnumeratable<KVP>` support for `ContainsKey(...).WithValue(...)`

### LINKED ISSUE #4852: Expand the usability of `Contains.Key(...).WithValue(...)`
I want to assert on HTTP response headers, but the `DictionaryContainsKeyValuePairConstraint` created by `Contains.Key(...).WithValue(...)` does not work on the type.

Currently the constraint checks for `IDictionary` and (despite the error message) also `IDictionary<TKey, TValue>`.

However, the `HttpHeaders` type implements `IEnumerable<KeyValuePair<string, IEnumerable<string>>>` (and of course the non-generic `IEnumerable` interface implicitly) and nothing else.

Would it be acceptable to expand the checks on this contraint to allow `IEnumerable<KeyValuePair<TKey, TValue>>` as well? I'm happy to work on it if the change would be accepted.

### PR DESCRIPTION
**Pull request by RenderMichael**
Fixes #4852

### CODE REVIEW

File: src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs [hunk 1]
```diff
@@ -70,6 +71,33 @@ private bool Matches(object? actual)
                 return false;
             }
 
+            // If 'actual' implements IEnumerable<KeyValuePair<TKey, TValue>>, we can find the value using that interface
+            var enumerableType = actual.GetType().GetInterface("IEnumerable`1");
```
Comment (manfred-brands) [inline src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs:75]: Suggest to cache the result of `actual.GetType().GetInterfaces()` from line 60.

File: src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs [hunk 2]
```diff
@@ -70,6 +71,33 @@ private bool Matches(object? actual)
...
+
+                    foreach (object item in (IEnumerable)actual)
+                    {
+                        var expectedKvp = new KeyValuePair<object, object?>(_expected.Key, _expected.Value);
+                        var actualKvp = new KeyValuePair<object, object?>(keyGetter.GetValue(item)!, valueGetter.GetValue(item));
+
+                        if (ItemsEqual(actualKvp, expectedKvp))
+                        {
+                            return true;
+                        }
```
Comment (manfred-brands) [inline src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs:94]: Is any of this needed? The code in the existing code above, even though it checks for `IDictionary` could as well only check for `IEnumerable<KeyValuePair>` as that is all that it uses.

**Review by manfred-brands — CHANGES_REQUESTED** [review]
The test that fails (`DictionaryContainsKeyValuePairConstraintTests.FailsWhenNotUsedAgainstADictionary`) is one that expects the code to fail if it isn't an `IDictionary`.
The test uses an List<KeyValuePair<,>>` which is the support you added.

Rewrite that test to pass or delete it as you already made a new test.

File: src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs (hunk 1 above)
Comment (manfred-brands) [reply src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs:75]: Not relevant if the original test on line 60 is changed to test for `IEnumerable<KeyValuePair<,>>` instead.

File: src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs (hunk 2 above)
Comment (RenderMichael) [reply src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs]: That's a good point. I changed the entire constraint to remove any dictionary-ness and instead make it work with any "key-value pair provider".

File: src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs [hunk 3]
```diff
@@ -41,7 +40,7 @@ public DictionaryContainsKeyValuePairConstraint(object key, object? value)
         private bool Matches(object? actual)
         {
             if (actual is null)
-                throw new ArgumentException("Expected: IDictionary But was: null", nameof(actual));
+                throw new ArgumentException("Expected: IEnumerable<KeyValuePair<,>> But was: null", nameof(actual));
```
Comment (manfred-brands) [inline src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs:43]: It also works for just `IDictionary`. 
So suggest to make it:
```suggestion
                throw new ArgumentException("Expected: IDictionary or IEnumerable<KeyValuePair<,>> But was: null", nameof(actual));
```

File: src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs [hunk 4]
```diff
@@ -70,7 +69,24 @@ private bool Matches(object? actual)
                 return false;
             }
 
-            throw new ArgumentException($"Expected: IDictionary But was: {actual.GetType()}", nameof(actual));
+            throw new ArgumentException($"Expected: IEnumerable<KeyValuePair<,>> But was: {actual.GetType()}", nameof(actual));
```
Comment (manfred-brands) [inline src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs:72]: Same here.
```suggestion
            throw new ArgumentException($"Expected type implementing IDictionary or IEnumerable<KeyValuePair<,>> But was: {actual.GetType()}", nameof(actual));
```

File: src/NUnitFramework/tests/Constraints/DictionaryContainsKeyValueConstraintTests.cs [hunk 5]
```diff
@@ -222,7 +221,24 @@ public void FailsWhenNullDictionary()
 
             TestDelegate act = () => Assert.That(dictionary, new DictionaryContainsKeyValuePairConstraint("1", "World"));
 
-            Assert.That(act, Throws.ArgumentException.With.Message.Contains("Expected: IDictionary But was: null"));
+            Assert.That(act, Throws.ArgumentException.With.Message.Contains("Expected: IEnumerable<KeyValuePair<,>> But was: null"));
```
Comment (manfred-brands) [inline src/NUnitFramework/tests/Constraints/DictionaryContainsKeyValueConstraintTests.cs:224]: Adjust to changes suggest above.

**Review by manfred-brands — CHANGES_REQUESTED** [review]
@RenderMichael Most of it looks good now.
Only item I have is the dropping of `IDictionary` in the exceptions.
Please put that back in.

**Comment by RenderMichael** [comment]
The error message has been fixed to refer to non-generic `IDictionary` again.

**Review by manfred-brands — APPROVED** [review]
@RenderMichael Thanks for your contribution. Looks good now.

**Pull request merged**